src/VisualStudio/Core/Def/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -278,26 +280,39 @@ public static WatcherOperation CombineRange(ImmutableSegmentedList<WatcherOperat
                 }
             }
 
-            return firstOp._kind switch
+            switch (firstOp._kind)
             {
-                Kind.WatchFiles =>
-                    WatchFiles(fileNamesBuilder.ToImmutable(), firstOp._fileChangeFlags, firstOp._sink, tokensBuilder.ToImmutable()),
-                Kind.UnwatchFiles =>
-                    UnwatchFiles(tokensBuilder.ToImmutable()),
-                Kind.UnwatchDirectories =>
-                    UnwatchDirectories(cookiesBuilder.ToList()),
-                _ =>
-                    throw ExceptionUtilities.Unreachable()
-            };
+                case Kind.WatchFiles:
+                    return WatchFiles(fileNamesBuilder.ToImmutable(), firstOp._fileChangeFlags, firstOp._sink, tokensBuilder.ToImmutable());
+                case Kind.UnwatchFiles:
+                    return UnwatchFiles(tokensBuilder.ToImmutable());
+                case Kind.WatchDirectory:
+                    filtersBuilder.RemoveDuplicates();
+                    return WatchDirectory(firstOp._paths[0], filtersBuilder.ToImmutable(), firstOp._sink, cookiesBuilder.ToList());
+                case Kind.UnwatchDirectories:
+                    return UnwatchDirectories(cookiesBuilder.ToList());
+            }
+
+            throw ExceptionUtilities.Unreachable();
         }
 
         public bool CanCombineWith(in WatcherOperation other)
         {
-            // Watching directory operation cannot be combined
-            if (_kind == Kind.WatchDirectory)
+            if (_kind != other._kind)
                 return false;
 
-            return (_kind == other._kind);
+            if (_kind == Kind.WatchFiles)
+            {
+                return other._sink == _sink;
```

Comment:
It might be easier/prudent just to compare the _sink regardless of the underlying type; that way we don't ever forget to update this check and combine a different kind of operation by accident.